### PR TITLE
Enhancement: add support to publishing helm chart 576

### LINKED ISSUE #576: Enhancement: add support to packaging and publishing helm chart
### Description
JKube doesn't support packaging and publishing  charts. It may be very useful to include those functionalities and so to avoid using additional maven plugin for this.

### Info

* If it's a feature request, what is your use case :
Add support for packaging and publishing helm chart.

### PR DESCRIPTION
**Pull request by dloiacono**
Signed-off-by: dloiacono <dloiacono@gmail.com>

## Description
Fix #576 
JKube doesn't support publishing charts. It may be very useful to include those functionalities and so to avoid using additional maven plugin for this.

## Type of change
 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [X] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [X] I have read the [contributing guidelines](https://www.eclipse.org/jkube/contributing)
 - [X] I signed-off my commit with a user that has signed the [Eclipse Contributor Agreement](https://www.eclipse.org/legal/ECA.php)
 - [X] I Added [CHANGELOG](../CHANGELOG.md) entry
 - [X] I have implemented unit tests to cover my changes
 - [X] I have updated the [documentation](../kubernetes-maven-plugin/doc) accordingly
 - [X] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=jkubeio_jkube) report
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

File: jkube-kit/jkube-kit-quarkus/src/main/java/org/eclipse/jkube/quarkus/generator/QuarkusGenerator.java [hunk 1]
```diff
@@ -145,7 +145,7 @@ private AssemblyConfiguration createAssemblyConfiguration(String targetDir, Asse
 
   private AssemblyFileSet getJvmFilesToInclude() {
     AssemblyFileSet.AssemblyFileSetBuilder fileSetBuilder =
-        getFileSetWithFileFromBuildThatEndsWith("-runner.jar");
+        getFileSetWithFileFromBuildThatEndsWith("-run.jar");
```
Comment (rohanKanojia) [inline jkube-kit/jkube-kit-quarkus/src/main/java/org/eclipse/jkube/quarkus/generator/QuarkusGenerator.java:148]: Umm, I think you're trying to modify QuarkusGenerator to add support for fast jar(https://github.com/eclipse/jkube/issues/632) . In my opinion, this change should go in a separate PR

Comment (dloiacono) [reply jkube-kit/jkube-kit-quarkus/src/main/java/org/eclipse/jkube/quarkus/generator/QuarkusGenerator.java]: I opened the https://github.com/eclipse/jkube/pull/638

File: jkube-kit/resource/helm/src/main/java/org/eclipse/jkube/kit/resource/helm/HelmUploader.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    String uploadUrl = repository.getUrl();
+    // Append slash if not already in place
+    if (!uploadUrl.endsWith("/")) {
+      uploadUrl += "/";
+    }
+    uploadUrl = uploadUrl + file.getName();
+
+    final HttpURLConnection connection = createConnection(uploadUrl);
+    connection.setDoOutput(true);
+    connection.setRequestMethod("PUT");
```
Comment (rohanKanojia) [inline jkube-kit/resource/helm/src/main/java/org/eclipse/jkube/kit/resource/helm/HelmUploader.java:130]: `getConnectionForUploadToNexus` and `getConnectionForUploadToArtifactory` look quite similar. Is it possible to move common code to some other method in order to avoid duplication?

File: jkube-kit/resource/helm/src/main/java/org/eclipse/jkube/kit/resource/helm/HelmUploader.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+      case NEXUS:
+        connection = getConnectionForUploadToNexus(file, repository);
+        break;
+      default:
+        throw new IllegalArgumentException("Unsupported repository type for upload.");
+    }
+
+    writeFileOnConnection(file, connection);
+
+    if (connection.getResponseCode() >= 300) {
```
Comment (rohanKanojia) [inline jkube-kit/resource/helm/src/main/java/org/eclipse/jkube/kit/resource/helm/HelmUploader.java:62]:
```suggestion
    if (connection.getResponseCode() >= HttpURLConnection.HTTP_MULT_CHOICE) {
```

File: kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/HelmPushMojo.java [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+    }
+    return getHelm().getStableRepository();
+  }
+
+  /**
+   * Get credentials for given helm repo. If username is not provided the repo
+   * name will be used to search for credentials in <code>settings.xml</code>.
+   *
+   * @param repository Helm repo with id and optional credentials.
+   * @return Authentication object or <code>null</code> if no credentials are present.
```
Comment (rohanKanojia) [inline kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/HelmPushMojo.java:132]:
```suggestion
```

File: kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/HelmPushMojo.java [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+   * @throws IllegalArgumentException Unable to get authentication because of misconfiguration.
+   */
+  void setAuthentication(HelmRepository repository)
+      throws IllegalArgumentException, SecDispatcherException {
```
Comment (rohanKanojia) [inline kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/HelmPushMojo.java:136]:
```suggestion
      throws SecDispatcherException {
```

**Review by rohanKanojia — APPROVED** [review]
Looks good with minor comments

**Comment by rohanKanojia** [comment]
@dloiacono: Hello, I wanted to test push goal. Which helm repository did you use to test the push goal? Is there some publicly available one just like docker hub?

**Comment by dloiacono** [comment]
> @dloiacono: Hello, I wanted to test push goal. Which helm repository did you use to test the push goal? Is there some publicly available one just like docker hub?

I used my company Artifactory

File: jkube-kit/resource/helm/src/main/java/org/eclipse/jkube/kit/resource/helm/HelmUploader.java (hunk 2 above)
Comment (dloiacono) [reply jkube-kit/resource/helm/src/main/java/org/eclipse/jkube/kit/resource/helm/HelmUploader.java]: I pushed some refactoring

**Comment by manusa** [comment]
**_WIP_**
How to test:

Deploy ChartMuseum:
```shell
 docker run --rm -it \
  -p 8080:8080 \
  -e DEBUG=1 \
  -e STORAGE=local \
  -e STORAGE_LOCAL_ROOTDIR=/charts \
  -e BASIC_AUTH_USER=user1 \
  -e BASIC_AUTH_PASS=s3cret \
  -e AUTH_ANONYMOUS_GET=true \
  -v $(pwd)/charts:/charts \
  ghcr.io/helm/chartmuseum:v0.13.1
```

**Comment by dloiacono** [comment]
@manusa You can also reuse the <server> configuration defined in the settings.xml by matching the repository name.

**Comment by manusa** [comment]
> @manusa You can also reuse the configuration defined in the settings.xml by matching the repository name.

:+1: thx, yes, I saw in the code. I think we're missing this in the documentation, but we can add it later. So far, it's looking very good.

I'm trying to check now with Nexus (since I think you already checked with Artifactory).

**Comment by dloiacono** [comment]
> > @manusa You can also reuse the configuration defined in the settings.xml by matching the repository name.
> 
> 👍 thx, yes, I saw in the code. I think we're missing this in the documentation, but we can add it later. So far, it's looking very good.
> 
> I'm trying to check now with Nexus (since I think you already checked with Artifactory).

Yes I already tested with Artifactory

**Review by manusa — APPROVED** [review]
LGTM, awesome job, thanks for contributing :heart:

**Comment by dloiacono** [comment]
@manusa I also tested on Artifactory, all is fine!
